src/polydatum/middleware.py

```diff
@@ -0,0 +1,184 @@
+    ):
+        self.ctx = ctx
+        self.path = path
+        self.args = args
+        self.kwargs = kwargs
+        self.dal_method = None  # Not resolved yet
+
+
+class DalMethodError(Exception):
+    """
+    This class defines a DalMethodError that is raised
+    when a service/method cannot be resolved by the
+    DalMethodResolverMiddleware
+    """
+    def __init__(self, path: Optional[Tuple[PathSegment]] = None):
+        self.path = path
+        super().__init__(f'Invalid dal method: {self.path}')
+
+
+class DeferredDalAttrRequest:
+    """
+    A utility class that defers access to DAL attributes.
+
+    Attribute access to the DAL is deferred so that the DAL can
+    run method middleware before and after calling a method.
+
+    Method middlewares may prevent method's from being called
+    by raising exceptions.
+    """
+
+    # Defining type here allows subclasses to easily provide another class
+    PathSegment = PathSegment
+
+    def __init__(self, handler: Callable, path: Tuple[PathSegment, ...]):
+        """
+        Args:
+            handler (Callable): A callable that handles calling the underlying
+                attribute by wrapping it so that method middleware can run before
+                and after calling a method
+            path (Tuple[PathSegment, ...]): A tuple of PathSegments representing
+                the current attribute path that has been requested.
+        """
+        self._handler = handler
+        self.path = path
+
+    def __getattr__(self, name: str) -> 'DeferredDalAttrRequest':
+        return self.__class__(self._handler, self.path + (self.PathSegment(name=name),))
+
+    def __call__(self, *args, **kwargs):
+        return self._handler(self.path, *args, **kwargs)
+
+    def __repr__(self):
+        location = [p.name for p in self.path]
+        location = '.'.join(location)
+        return f"<{self.__class__} paths={location}>"
+
+
+def dal_resolver(ctx, deferred_dal_attr_request):
+    """
+    This function resolves a dal method call to its underlying
+    service
+    """
+    service_or_method = None
+    for (_path, _service_or_method) in resolve(ctx, deferred_dal_attr_request.path):
+        if not _service_or_method:
+            raise DalMethodError(path=tuple(_path))
+        elif not isinstance(_service_or_method, (Callable, Service)):
+            raise DalMethodError(path=tuple(_path))
+        service_or_method = _service_or_method
+    return service_or_method
+
+
+def resolve(ctx: DataAccessContext, path: Tuple[PathSegment]):
+    """
+    A generator that returns a list of path segments traversed
+    during resolving the dal attribute and the method to call, if any.
+    """
+    service_or_method = ctx.dal._services  # noqa
+    paths = list(path)
+    location = []
+
+    while paths:
+        path_segment = paths.pop(0)
+        location.append(path_segment)
```

Comment:
`location += (path_segment,)` would allow this to be a `tuple` which is how path is used in other places. I wonder what the performance difference of a list vs a tuple is here.